src/cascadia/TerminalSettingsModel/ActionMap.cpp

```diff
@@ -613,76 +561,28 @@ namespace winrt::Microsoft::Terminal::Settings::Model::implementation
-                _MaskingActions.emplace(conflictingActionID, *conflictingCmdCopy);
-            }
-            else
-            {
-                // We've collided with this action before. Let's resolve a collision with a masking action.
-                const auto maskingCmdImpl{ get_self<implementation::Command>(maskingCmdPair->second) };
-                maskingCmdImpl->EraseKey(keys);
-            }
-        }
+        if (const auto foundCommand = _GetActionByKeyChordInternal(keys); foundCommand && *foundCommand)
+        {
+            // collision: the key chord is bound to some command, make sure that command erases
+            //            this key chord as we are about to overwrite it
 
-        // Assign the new action in the _KeyMap.
-        const auto actionID{ Hash(cmd.ActionAndArgs()) };
-        _KeyMap.insert_or_assign(keys, actionID);
+            const auto foundCommandImpl{ get_self<implementation::Command>(*foundCommand) };
+            foundCommandImpl->EraseKey(keys);
+        }
 
-        // Additive operation:
-        //   Register the new key chord with oldCmd (an existing _ActionMap entry)
-        // Example:
-        //   { "command": "copy",  "keys": "ctrl+c" }       --> register "ctrl+c" (section above)
-        //   { "command": "copy",  "keys": "ctrl+shift+c" } --> also register "ctrl+shift+c" to the same Command (oldCmd)
-        if (oldCmd)
+        // Assign the new action in the _KeyMap
+        // However, there's a strange edge case here - since we're parsing a legacy or modern block,
+        // the user might have { "command": null, "id": "someID", "keys": "ctrl+c" }
+        // i.e. they provided an ID for a null command (which they really shouldn't, there's no purpose)
+        // in this case, we do _not_ want to use the id they provided, we want to use an empty id
+        // (empty id in the _KeyMap indicates the keychord was explicitly unbound)
+        if (cmd.ActionAndArgs().Action() == ShortcutAction::Invalid)
         {
-            // Update inner Command with new key chord
-            auto oldCmdImpl{ get_self<Command>(oldCmd) };
-            oldCmdImpl->RegisterKey(keys);
+            _KeyMap.insert_or_assign(keys, L"");
         }
-
-        // Additive operation:
-        //   Register the new key chord with maskingCmd (an existing _maskingAction entry)
-        // Example:
-        //   parent:  { "command": "copy", "keys": "ctrl+c" }       --> register "ctrl+c" to parent._ActionMap (different branch in a different layer)
-        //   current: { "command": "copy", "keys": "ctrl+shift+c" } --> also register "ctrl+shift+c" to the same Command (maskingCmd)
-        if (maskingCmd)
+        else
         {
-            // Update inner Command with new key chord
-            auto maskingCmdImpl{ get_self<Command>(maskingCmd) };
-            maskingCmdImpl->RegisterKey(keys);
+            _KeyMap.insert_or_assign(keys, cmd.ID());
         }

```

Comment:
This could be something like
```cpp
const auto action = cmd.ActionAndArgs().Action();
const auto id = action == ShortcutAction::Invalid ? hstring{} : cmd.ID();
_KeyMap.insert_or_assign(keys, id);
```

I'm not sure I understand why we want to store an empty string in the map for invalid actions though. Shouldn't we want to remove the key entirely?

Reply:
So we can tell when a keychord was explicitly unbound vs simply unused